autoconfigure/src/main/java/info/developerblog/spring/cloud/marathon/discovery/MarathonDiscoveryClient.java

```diff
@@ -44,73 +50,81 @@ public ServiceInstance getLocalServiceInstance() {
-             */
-                try {
-                    GetAppResponse response = client.getApp(ServiceIdConverter.convertToMarathonId(serviceId));
-
-                    if (response != null && response.getApp() != null)
-                        instances.addAll(extractServiceInstances(response.getApp()));
-
+        	List<ServiceInstance> instances = new ArrayList<>();
+        	
+        	if (!allServices(serviceId)) {
+        		try {
+        			instances.addAll(getInstance(serviceId));
                 } catch (MarathonException e) {
                     log.error(e.getMessage(), e);
                 }
-            }
-
-            if (instances.isEmpty()) {
-
-                /*
-                Step 2 - Search for all applications whose marathon id contains the service id (e.g. "*.{serviceId}*.")
-                This is supported by the marathon api by passing a partial id as a query parameter
-                 */
-                Map<String,String> queryMap = new HashMap<>();
-                queryMap.put("id",ServiceIdConverter.convertToMarathonId(serviceId));
-
-                GetAppsResponse appsResponse = (ignoreServiceId)?client.getApps():client.getApps(queryMap);
-
-                if (appsResponse!=null && appsResponse.getApps()!=null) {
-
-                	List<App> apps = appsResponse.getApps();
-                	
-                	log.debug("Discovered {} service{}{}", apps.size(), apps.size() == 1 ? "" : "s", ignoreServiceId ? "" : String.format(" with ids that contain [%s]", serviceId));
-
-                    for (App app : apps){
-
-                        // Fetch data for this specific service id, to collect task information
-                        GetAppResponse response = client.getApp(app.getId());
-
-                        if (response!=null && response.getApp()!=null)
-                            instances.addAll(extractServiceInstances(response.getApp()));
-
-                    }
-                }
-            }
-
-            log.debug("Discovered {} service{}{}", instances.size(), instances.size() == 1 ? "" : "s", ignoreServiceId ? "" : String.format(" with ids that contain [%s]", serviceId));
+        		
+        		if (instances.isEmpty()) {
+        			instances.addAll(getInstances(new SearchServiceQuery(serviceId)));
+        		}
+        	} else {
+        		instances.addAll(getInstances());
+        	}
+        	
+            log.debug("Discovered {} service{}{}", instances.size(), instances.size() == 1 ? "" : "s", allServices(serviceId) ? "" : String.format(" with ids that contain [%s]", serviceId));
             return instances;
 
         } catch (Exception e) {
             log.error(e.getMessage(), e);
             return Collections.emptyList();
         }
     }
-
-
-    /**
+    
+    private List<ServiceInstance> getInstances() throws MarathonException {
+    	return getInstances((SearchServiceQuery)null);
```

Comment:
We can to hide the selection of search of some or all instances into Query object which will be responsible of calling of getApps method